### PR TITLE
Savani initial condition

### PR DESCRIPTION
**Pull request by Shangxin-Liu**
Since SAVANI and S40RTS use the same interpolation methods between different spherical layers, I move the cubic spline interpolation (namespace tk) to the utilities (definition part in the utilities.h, implementation part in the utilities.cc). If in the future other tomography models need to use this method, it will be quite convenient to call the namespace. If not, we can add other interpolation methods in the utilities. I also attach the test results files here. They all look good. SAVANI test actually cannot take that long time (I made a mistake in the Hackathon). I also have successfully used SAVANI as the initial condition to generate a reasonable global dynamic topography map similar but different from that of S40RTS.

### CODE REVIEW

File: tests/initial_condition_SAVANI/screen-output [hunk 1]
```diff
@@ -0,0 +1,39 @@
+-----------------------------------------------------------------------------
```
Comment (bangerth) [inline tests/initial_condition_SAVANI/screen-output]: You forgot to add the .prm file for this test.

File: include/aspect/initial_conditions/S40RTS_perturbation.h [hunk 2]
```diff
@@ -31,7 +31,7 @@ namespace aspect
   {
     using namespace dealii;
 
-    namespace internal
+    namespace S40RTSinternal
```
Comment (bangerth) [inline include/aspect/initial_conditions/S40RTS_perturbation.h]: Why is this change necessary? Do you get linker errors because you declare the same classes in your own model?

In that case, my preference would be to use
```
namespace internal
{
  namespace S40RTS
  {
    class SphericalHarmonics;
```
and similar, but with a different *inner* namespace in your own model.

Of course, this begs the question: are the classes you declare actually different and just happen to have the same name? Or do you simply use a copy?

**Comment by bangerth** [comment]
I'm glad to hear you got the compute time down to something more reasonable. Since we can all learn from mistakes, would you share what your mistake was?

File: tests/initial_condition_SAVANI/screen-output (hunk 1 above)
Comment (Shangxin-Liu) [inline tests/initial_condition_SAVANI/screen-output]: Oh, I omitted it. Here it is.

File: include/aspect/initial_conditions/S40RTS_perturbation.h (hunk 2 above)
Comment (Shangxin-Liu) [inline include/aspect/initial_conditions/S40RTS_perturbation.h]: I found that S40RTS and SAVANI cannot have the same namespace "internal", that's why in the Hackathon the SAVANI test cannot finished with 600s. It seems they got some "mess-up" problem if they use the same namespace. S40RTSinternal and SAVANIinternal are similar, but not the same, because their data structures are different. In addition to the different spherical harmonic coefficients, the center interpolation depth and the number of layers are also different (SAVANI-28, S40RTS-21). I once thought I can add if loop in the code to include two initial conditions, but I suppose it's not so elegant- One source file containing two initial condition. Also, in the future, what if we add other tomography model using different interpolation methods? I think InitialConditions::S40RTSinternal/SAVANIinternal seems good. Or can we replace S40RTSinternal directly with S40RTS and the same with SAVANI?

Comment (bangerth) [inline include/aspect/initial_conditions/S40RTS_perturbation.h]: You can't just use the names `S40RTS` because that's already the name of the class.

Just go with
```
namespace internal
{
  namespace S40RTS
  {
    class SphericalHarmonics;
```
This is the style we use everywhere else already.

**Comment by Shangxin-Liu** [comment]
According to the tests, the mistake in the Hackathon is that I use the same namespace internal in the two initial conditions so the test cannot finish and the computed results are not correct. This test with shorter compute time has nothing to do with the slow convergence problem in the variable viscosity. In the lower global refinement degree, all cases can converge at the default Tol of 1e-7. But with higher global refinement degree (>2), the slow (or not) convergence problem still appears. I suppose the authors of all kinds of material models maybe just performed tests with refinement degree 0 so the convergence problem doesn't appear. 

We're now doing some benchmark cases with the basic harmonic perturbation initial conditions between ASPECT and CitcomS to see the run time and convergence. And we have precluded the possibility that the slow run time is due to quadrature points compute algorithms of S40RST/SAVANI, because in the higher global refinement cases, the solver accounts for the most of the run time.

File: include/aspect/initial_conditions/S40RTS_perturbation.h (hunk 2 above)
Comment (Shangxin-Liu) [inline include/aspect/initial_conditions/S40RTS_perturbation.h]: OK. I'll fix this.

Comment (Shangxin-Liu) [inline include/aspect/initial_conditions/S40RTS_perturbation.h]: Just make sure. So I'll adjust the code to aspect::InitialConditions::internal::S40RTS?

Comment (bangerth) [inline include/aspect/initial_conditions/S40RTS_perturbation.h]: correct.

**Comment by Shangxin-Liu** [comment]
Here are the new S40RTS/SAVANI files. I've run the test and got the same results as before. @bangerth

**Comment by bangerth** [comment]
Thanks, @Shangxin-Liu . You have a lot of whitespace changes in your patch that trip up our indentation script, see http://www.math.clemson.edu/~heister/aspect-astyler-logs/1da0025a4d8f0e646f8b5c78d78b6467e78b13b4/log . Can you run astyle (or doc/modules/indent) on your new files?

**Comment by Shangxin-Liu** [comment]
I've run the indent script for all files. While I'm making the new commit for all related files including the test results, I found that the indent fix only changes two files a bit. Since there is an new namespace for each of the four files, I suppose this may induce quite a bit indent difference from the previous ones. @bangerth

**Comment by Shangxin-Liu** [comment]
It seems that the previous commits are influenced by some tmp files with a "~" end. I recommit everything in a new clean directory, the utilities files, SAVANI/S40RTS files, SAVANI data files, and test results files. Now the astyle-tester seems good. @bangerth

**Comment by Shangxin-Liu** [comment]
The test results are in the second "squashed one" commit, the last three files.@tjhei

**Comment by bangerth** [comment]
Thanks, @Shangxin-Liu ! In a follow-up commit, can you please add an entry to `doc/modules/changes.h` that announces the new plugin?

**Comment by Shangxin-Liu** [comment]
Is there any way to add one extra commit to a merged pull request? It seems I have to create a new one for the new entry to change.h. @bangerth

**Comment by bangerth** [comment]
No, you'll just have to create a new pull request. I should have thought of that before hitting the merge button.

**Comment by Shangxin-Liu** [comment]
OK. I've created a new one.
